runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/WindmillStateReaderTest.java

```diff
@@ -99,6 +113,713 @@ private ByteString intData(int value) throws IOException {
+            STATE_MULTIMAP_KEY_1, STATE_KEY_1, STATE_FAMILY, INT_CODER);
+    Future<Iterable<Integer>> future2 =
+        underTest.multimapFetchSingleEntryFuture(
+            STATE_MULTIMAP_KEY_2, STATE_KEY_1, STATE_FAMILY, INT_CODER);
+    Mockito.verifyNoMoreInteractions(mockWindmill);
+
+    Windmill.KeyedGetDataRequest.Builder expectedRequest =
+        Windmill.KeyedGetDataRequest.newBuilder()
+            .setKey(DATA_KEY)
+            .setShardingKey(SHARDING_KEY)
+            .setWorkToken(WORK_TOKEN)
+            .setMaxBytes(WindmillStateReader.MAX_KEY_BYTES)
+            .addMultimapsToFetch(
+                Windmill.TagMultimapFetchRequest.newBuilder()
+                    .setTag(STATE_KEY_1)
+                    .setStateFamily(STATE_FAMILY)
+                    .setFetchEntryNamesOnly(false)
+                    .addEntriesToFetch(
+                        Windmill.TagMultimapEntry.newBuilder()
+                            .setEntryName(STATE_MULTIMAP_KEY_1)
+                            .setFetchMaxBytes(WindmillStateReader.INITIAL_MAX_MULTIMAP_BYTES)
+                            .build())
+                    .addEntriesToFetch(
+                        Windmill.TagMultimapEntry.newBuilder()
+                            .setEntryName(STATE_MULTIMAP_KEY_2)
+                            .setFetchMaxBytes(WindmillStateReader.INITIAL_MAX_MULTIMAP_BYTES)
+                            .build()));
+
+    Windmill.KeyedGetDataResponse.Builder response =
+        Windmill.KeyedGetDataResponse.newBuilder()
+            .setKey(DATA_KEY)
+            .addTagMultimaps(
+                Windmill.TagMultimapFetchResponse.newBuilder()
+                    .setTag(STATE_KEY_1)
+                    .setStateFamily(STATE_FAMILY)
+                    .addEntries(
+                        Windmill.TagMultimapEntry.newBuilder()
+                            .setEntryName(STATE_MULTIMAP_KEY_1)
+                            .addAllValues(Arrays.asList(intData(5), intData(6))))
+                    .addEntries(
+                        Windmill.TagMultimapEntry.newBuilder()
+                            .setEntryName(STATE_MULTIMAP_KEY_2)
+                            .addAllValues(Arrays.asList(intData(15), intData(16)))));
+    when(mockWindmill.getStateData(ArgumentMatchers.eq(COMPUTATION), ArgumentMatchers.any()))
+        .thenReturn(response.build());
+
+    Iterable<Integer> results1 = future1.get();
+    Iterable<Integer> results2 = future2.get();
+
+    final ArgumentCaptor<Windmill.KeyedGetDataRequest> requestCaptor =
+        ArgumentCaptor.forClass(Windmill.KeyedGetDataRequest.class);
+    Mockito.verify(mockWindmill)
+        .getStateData(ArgumentMatchers.eq(COMPUTATION), requestCaptor.capture());
+    assertMultimapFetchRequestEqual(
+        expectedRequest.build().getMultimapsToFetch(0),
+        requestCaptor.getValue().getMultimapsToFetch(0));
+
+    // Iterate over the results to force loading all the pages.
```

Comment:
nit: could you just verifyNoMoreInteractions after doing assertThat(results1, ...) and have that do the iteration?